.github/pull_request_template.md

```diff
@@ -0,0 +1,53 @@
+Fixes #XXXX.
+
+(Replace XXXX with the id of the issue fixed in this PR. Remove this line if there is no corresponding
+issue. Don't reference the issue in the title of this pull-request.)
+
+Add tags to your PR if you are a committer (only committers have the right to add tags). Add [Design Review] tag
+if this PR should better be reviewed by at least two people.
+Don't forget to add the following tags (if applicable): [Incompatible], [Release Notes], [Compatibility], [Security],
+[Development Blocker]. Add at least one [Area - ] tag, consider creating a new one if none of the existing [Area - ]
+tags is applicable.
+
+### Description
+
+Describe the goal of this PR, what problem are you fixing. If there is a corresponding issue (referenced above), it's
+not necessary to repeat the description here, however, you may choose to keep one summary sentence.
+
+Describe your patch: what did you change in code? How did you fix the problem?
+
+If there are several relatively logically separate changes in this PR, list them. For example:
+ - Fixed the bug ...
+ - Renamed the class ...
+ - Added a forbidden-apis entry ...
+
+Some of the aspects mentioned above may be omitted for simple and small PRs.
+
+### Design
```

Comment:
I don't understand why this section exists. As far as I understand, every PR requiring design review should have a proposal according to the [discussion on dev](https://lists.apache.org/thread.html/bb9c5e1f8ce9b3148a5c26f95059f9b6629fae3bf8c617121d671395@%3Cdev.druid.apache.org%3E). Am I misunderstood?

Reply:
Every code change has a design. Even choosing a name for a single variable is already an act of design. This section shouldn't discuss variable naming (although sometimes it can), but it can involve the following sentences:
- "This new query returns errors in cases of X, but returns empty results in case of Y, because ..."
- "I've used the factory (see MyNewCoolFactory) pattern instead of the builder pattern, because ..."

Real examples: in [this change](https://github.com/apache/incubator-druid/pull/7185), I could (and probably should) have included the text from [this comment](https://github.com/apache/incubator-druid/pull/7185#discussion_r264413165) into the Design section of the PR, because it's an act of naming design.

@egor-ryashin's #6349 is a good example, because the whole PR's description is actually description of the design of the change. Not only code design, but also *the design of the solution*. Where the feature is "we want to be able to decommission nodes safely", the design of the solution is "we provide a list of nodes to decommission to Coordinator via its dynamic config". This is very important and should always be described in the PR description.

What's also good about that description is that it presents the *alternative design* and compares it with the chosen design.

P. S. after writing this comment it seems to me that the sentence
```
Describe your patch: what did you change in code? How did you fix the problem?
```
From the "Description" section above should actually belong to the "Design" section.

Reply:
There's a lot of overlap between this "Design" and the definition of a proposal: https://github.com/apache/incubator-druid/blob/master/.github/ISSUE_TEMPLATE/proposal.md. It's weird to have that template and this one talking about the same things from different angles. One way to address that is to take anything useful from this template that isn't already in the Proposal template, then merge that stuff into the Proposal template and simply reference that one here. And then open this template by saying that:

1. If you raised a proposal, just link to it.
2. If you didn't raise a proposal, think about doing that first, especially if the PR has any sort of 'interestingness' to its design that is worth discussing. If not, then include the same kind of information here in the PR.

I feel that encouraging people to raise proposals is useful because it promotes separation of design-level and implementation-level discussion. One point that was raised in the dev list when we talked about this is that commingling them can cause one to drown out the other.

Reply:
@gianm's suggestion sounds good to me. Different aspects of design in different templates looks ambiguous.

Reply:
I feel there is a fundamental problem that is behind this "separate Proposal issue" topic: we cannot really foresee what we will implement until we actually implement it. The final result may be drastically different from what we planned. The performance data may lead us to rethink the feature or its scope. The discussion in the Proposal may become irrelevant.

It's impossible to separate "design" and "implementation" (and therefore their discussions), neither in code, nor in time, nor in ideas.

This is why I think the PR template should contain the design section as well. 

Maybe instead of
```
If you already did this in the associated issue
(e. g. a "Proposal" issue), leave the following sentence:

 Design of this change is discussed [here](<link to Github 
issue or comment where you discuss the design>).
```

at the end of the section we can add:

```
If there was a discussion of the feature implemented in this PR
elsewhere (e. g. a "Proposal" issue, any other issue, or a thread
in the development mailing list), link to that discussion and
explain what did change in your final design compared to your
original proposal or the consensus version in the end of the
discussion. If something hasn't changed since the original discussion,
you can omit a detailed discussion of those aspects of the design here,
perhaps apart from brief mentioning for the sake of readability of
this PR description.
```

It doesn't seem any practical to me to suggest creating separate issues for design discussions after implementing the whole change. Was it discussed?

Reply:
I think there is value in separating them for PRs where a meaningful design discussion, separate from the code-level discussion, is anticipated. (Conversely: if no meaningful discussion is anticipated outside of the code level, then I don't see a point in separating the proposal from the PR.)

This is an example of one where I am finding it difficult to follow the design- and code-level discussions that are happening concurrently in the same PR: https://github.com/apache/incubator-druid/pull/7306. I think it would have benefited from a design proposal separate from the code-level changes.

Reply:
Another way of putting this is that code-level discussions tend to get into the weeds pretty quickly, and a lot of code-level back and forth can too easily obscure a larger discussion about overall approach and design.

Reply:
Does this mean that you disagree with the current wording of the template?

Reply:
I think the "Design" section is good in the current draft. I do think we could use some cleaning up of our community guidelines around when a proposal is and isn't needed. But that doesn't need to be done as part of this PR to create a template.

I asked https://github.com/apache/incubator-druid/pull/7206#issuecomment-485529475 mainly because of the two checklists (general & concurrency), which I think need to be clarified as to whether it's meant to be expressing requirements or suggestions. My preference would be 'suggestions' because I think it's best if the list of requirements is kept short, and the concurrency checklist is quite long.